File: 01-week-1--starter-algorithms/01-day-2--find-first-duplicate/javascript/find_first_duplicate.js
function findFirstDuplicate(arr) {
  // create Set object
  let elementSet = new Set();

  // loop through array and compare in Set object
  // if match, return result, else, add it to the Set
  for (let i = 0; i < arr.length; i++) {
    if (elementSet.has(arr[i])) return arr[i];
    elementSet.add(arr[i]);
  }

  // if no duplicates, return -1
  return -1;
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 3");
  console.log("=>", findFirstDuplicate([2, 1, 3, 3, 2]));

  console.log("");

  console.log("Expecting: 'add'");
  console.log("=>", findFirstDuplicate(['first', 'add', 'second', 'add', 'third']));

  console.log("");

  console.log("Expecting: -1");
  console.log("=>", findFirstDuplicate([1, 2, 3, 4]));
}

module.exports = findFirstDuplicate;

// Please add your pseudocode to this file
// And a written explanation of your solution

File: 01-week-1--starter-algorithms/01-day-2--find-first-duplicate/ruby/find_first_duplicate.rb
def find_first_duplicate(arr)
  # create counter hash with default 0
  counter_hash = Hash.new(0)

  # iterate over each element of array
  arr.each do |e|
    counter_hash[e] = counter_hash[e] += 1

    # if the counter_hash key reaches 2
    return e if counter_hash[e] > 1
  end

  # if no matches
  return -1
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 3"
  puts "=>", find_first_duplicate([2, 1, 3, 3, 2])

  puts

  puts "Expecting: -1"
  puts "=>", find_first_duplicate([1, 2, 3, 4])

  # Don't forget to add your own!
  puts

  puts "Expecting: 'added'"
  puts "=>", find_first_duplicate(['first', 'added', 'second', 'added', 'third'])

  puts 

  puts "Expecting: -1"
  puts "=>", find_first_duplicate([])
end

# Please add your pseudocode to this file
# And a written explanation of your solution
